### PR TITLE
Add support for AES-GCM and ChaCha20Poly1305 on iOS/tvOS/MacCatalyst via CryptoKit

### LINKED ISSUE #52482: ChaCha20Poly1305.IsSupported always returns false on non-Windows
Follow-up to https://github.com/dotnet/runtime/issues/45130. We checked in an implementation of ChaCha20Poly1305 for Windows, but we don't have one for Mac / Linux scenarios.

This work item tracks calling in to the existing CryptoKit / OpenSSL implementation for Mac / Linux scenarios so that we don't forget to do this work.

Implementations needed for:

- [x] Unix (https://github.com/dotnet/runtime/pull/52522)
- [x] Android (https://github.com/dotnet/runtime/pull/52814)
- [ ] iOS
- [x] Mac OS (https://github.com/dotnet/runtime/pull/76317)

Marked up for grabs.

### LINKED ISSUE #91523: Support AES-GCM for iOS-like platforms
[AES-GCM algorithm](https://learn.microsoft.com/en-us/dotnet/api/system.security.cryptography.aesgcm) (as well as [ChaCha20Poly1305](https://learn.microsoft.com/en-us/dotnet/api/system.security.cryptography.chacha20poly1305) and [AES-CCM](https://learn.microsoft.com/en-us/dotnet/api/system.security.cryptography.aesccm)) are available for macOS, [powered by CryptoKit](https://github.com/dotnet/runtime/issues/29811), but not on iOS.

As [CryptoKit](https://developer.apple.com/documentation/cryptokit/) is also available on iOS (iOS 13+), maybe there is an easy way to also bring support to iOS?

Motivation: I am building a cross-platform app for Windows, iOS, Android & macOS. Missing support for AES-GCM on iOS forces me to either pull in an additional dependency (e.g. BouncyCastle) or use AES-CBC (which seemingly the whole internet is advising against in favor of AES-GCM)

NOTE: once this is implemented we should update the docs: https://github.com/dotnet/docs/blob/main/docs/standard/security/cross-platform-cryptography.md

### PR DESCRIPTION
**Pull request by akoeplinger**
Now that we support minimum OS versions that ship with Swift we can enable CryptoKit and pal_swiftbindings.swift.

Fixes https://github.com/dotnet/runtime/issues/91523
Fixes https://github.com/dotnet/runtime/issues/52482

### CODE REVIEW

File: src/native/libs/System.Security.Cryptography.Native.Apple/CMakeLists.txt [hunk 1]
```diff
@@ -42,16 +42,38 @@ endif()
...
-    set(SWIFT_PLATFORM_SUFFIX "")
+if (NOT SWIFT_COMPILER_TARGET)
     set(SWIFT_DEPLOYMENT_TARGET ${CMAKE_OSX_DEPLOYMENT_TARGET})
+    set(SWIFT_PLATFORM_SUFFIX "")
+
+    if (CMAKE_BUILD_TYPE STREQUAL "Debug")
+        set(SWIFT_OPTIMIZATION_FLAG "-Onone")
+    elseif (CMAKE_BUILD_TYPE STREQUAL "Release")
+        set(SWIFT_OPTIMIZATION_FLAG "-O")
+    endif()
```
Comment (akoeplinger) [inline src/native/libs/System.Security.Cryptography.Native.Apple/CMakeLists.txt:53]: I noticed that we didn't compile the swift code with optimizations so fixed this as well

File: src/libraries/Common/tests/System/Security/Cryptography/AlgorithmImplementations/ECDsa/ECDsaTests.netcoreapp.cs [hunk 2]
```diff
@@ -188,7 +188,6 @@ public void KeySizeProp()
         }
 
         [Theory, MemberData(nameof(TestNewCurves))]
-        [ActiveIssue("https://github.com/dotnet/runtime/issues/51332", TestPlatforms.iOS | TestPlatforms.tvOS | TestPlatforms.MacCatalyst)]
```
Comment (vcsjones) [inline src/libraries/Common/tests/System/Security/Cryptography/AlgorithmImplementations/ECDsa/ECDsaTests.netcoreapp.cs:191]: This seems unrelated from AES-GCM and ChaCha20Poly1305. Does this test pass on those platforms?

File: src/libraries/System.Security.Cryptography/tests/AesCcmTests.cs [hunk 3]
```diff
@@ -12,7 +12,6 @@ namespace System.Security.Cryptography.Tests
     public class AesCcmTests : CommonAEADTests
     {
         [Theory]
-        [ActiveIssue("https://github.com/dotnet/runtime/issues/51332", TestPlatforms.iOS | TestPlatforms.tvOS | TestPlatforms.MacCatalyst)]
```
Comment (vcsjones) [inline src/libraries/System.Security.Cryptography/tests/AesCcmTests.cs:15]: This is for CCM, not GCM. I would not expect these tests to pass on mobile apple crypto.

File: src/native/libs/System.Security.Cryptography.Native.Apple/extra_libs.cmake [hunk 4]
```diff
@@ -5,9 +5,11 @@ macro(append_extra_cryptography_apple_libs NativeLibsExtra)
 
     list(APPEND ${NativeLibsExtra} ${COREFOUNDATION_LIBRARY} ${SECURITY_LIBRARY})
 
-    if (CLR_CMAKE_TARGET_OSX)
+    if (CLR_CMAKE_TARGET_OSX OR CLR_CMAKE_TARGET_MACCATALYST OR CLR_CMAKE_TARGET_IOS OR CLR_CMAKE_TARGET_TVOS)
```
Comment (vcsjones) [inline src/native/libs/System.Security.Cryptography.Native.Apple/extra_libs.cmake:8]: According to https://github.com/dotnet/runtime/issues/91736, the minimum iOS / tvOS target is 12.2. But we are strongly linking against CryptoKit which is not available until 13.0. What is the behavior of this? Will an App on 12.2 fail to launch because it can't find CryptoKit?

Comment (filipnavara) [reply src/native/libs/System.Security.Cryptography.Native.Apple/extra_libs.cmake:8]: Are we strongly linking into CryptoKit though? With Clang set to a specific min OS version the compiler uses the availability attributes in C headers, and the "text-based stub" (.tbd) linker files to automatically produce weak links. I would expect Swift to behave similarly.

Comment (vcsjones) [reply src/native/libs/System.Security.Cryptography.Native.Apple/extra_libs.cmake:8]: That could be. I am recalling though that folks on macOS 10.14 got app startup problems though on macOS (see https://github.com/dotnet/runtime/issues/93167 for an example). If it were weakly linked, I would have expected the app to work (even if it was on an unsupported OS).

Even if it is weakly linked, I wonder if our shims too-eagerly try to do something with it. Again using the previous example, that person was trying to use MD5, but they got CryptoKit errors.

Comment (filipnavara) [reply src/native/libs/System.Security.Cryptography.Native.Apple/extra_libs.cmake:8]: AFAIK #93167 happened after the Clang switch for minimum version was already bumped to 10.15. That automatically turns the weak links into strong ones since the library is part of the minimum target.

File: src/libraries/System.Security.Cryptography/ref/System.Security.Cryptography.cs [hunk 5]
```diff
@@ -121,6 +121,9 @@ public sealed partial class AesCryptoServiceProvider : System.Security.Cryptogra
     [System.Runtime.Versioning.UnsupportedOSPlatformAttribute("browser")]
     [System.Runtime.Versioning.UnsupportedOSPlatformAttribute("ios")]
     [System.Runtime.Versioning.UnsupportedOSPlatformAttribute("tvos")]
+    [System.Runtime.Versioning.SupportedOSPlatform("maccatalyst")]
+    [System.Runtime.Versioning.SupportedOSPlatform("ios13.0")]
+    [System.Runtime.Versioning.SupportedOSPlatform("tvos13.0")]
```
Comment (bartonjs) [inline src/libraries/System.Security.Cryptography/ref/System.Security.Cryptography.cs:126]: I'm so glad we stopped using SupportedOS on newer classes.

File: src/libraries/System.Security.Cryptography/tests/AesGcmTests.cs [hunk 6]
```diff
@@ -1005,14 +987,10 @@ public static void CheckIsSupported()
         {
             bool expectedIsSupported = !PlatformDetection.IsBrowser;
 
-            if (PlatformDetection.IsOSX)
+            if (PlatformDetection.IsOSX || PlatformDetection.UsesMobileAppleCrypto)
```
Comment (bartonjs) [inline src/libraries/System.Security.Cryptography/tests/AesGcmTests.cs:990]: Looking at the body of this function now, this if can just be deleted.  Neither macOS nor iOS/tvOS are browser, so it's already true, and it's not setting up an exclusion filter (in case IsOSX also implied UsesMobileAppleCrypto, which seems weird, but might be true).  So now it's

```C#
bool expect = !Browser;

if (ASpecificNotBrowser || ADifferentNotBrowser)
{
    expect = true;
}
```

which reduces to just `bool expect = !IsBrowser;`

File: src/native/libs/System.Security.Cryptography.Native.Apple/CMakeLists.txt [hunk 7]
```diff
@@ -42,16 +42,38 @@ endif()
...
+
+    if (CLR_CMAKE_TARGET_OSX)
+        set(SWIFT_PLATFORM "macosx")
+    elseif (CLR_CMAKE_TARGET_MACCATALYST)
+        set(SWIFT_PLATFORM "ios")
+        set(SWIFT_PLATFORM_SUFFIX "-macabi")
+    elseif (CLR_CMAKE_TARGET_IOS)
+        set(SWIFT_PLATFORM "ios")
+        if(CMAKE_OSX_SYSROOT MATCHES iPhoneSimulator)
```
Comment (bartonjs) [inline src/native/libs/System.Security.Cryptography.Native.Apple/CMakeLists.txt:62]:
```suggestion
        if (CMAKE_OSX_SYSROOT MATCHES iPhoneSimulator)
```

All the others ifs have a space after them.

File: src/native/libs/System.Security.Cryptography.Native.Apple/extra_libs.cmake (hunk 4 above)
Comment (filipnavara) [reply src/native/libs/System.Security.Cryptography.Native.Apple/extra_libs.cmake:8]: Just a refresher on how it works:
1. The system headers have annotations for each API which says on which platform+version it's available
2. C compiler takes the `-m[macosx/ios/...]-version-min` switch which specifies the platform+version you are targeting
3. If there's mismatch between the annotation, usage and the minimum version, the C compiler with emit a warning/error
4. Conditional `if (__builtin_available(...))` block can be used to check for specific platform/version and avoid the warning/error by telling the compiler that the code can only run on given platform. This emits either a no-op or runtime check depending on the intersection with specified minimum supported platform version. Alternatively, an availability attribute can be place on the calling method and thus move the check requirement up the call chain.
5. Linker has import libraries/frameworks specified as .tbd files that include the minimal platform version. Based on intersection with the `-m[macosx/ios/...]-version-min` switch it generates either weak or strong imports.

.NET equivalents:
1. `SupportedOSPlatform` attribute
2. `SupportedOSPlatformVersion` property and TFM in .csproj
3. Platform Compatibility Analyzer running inside Roslyn (part of .NET SDK) will emit a warning/error
4. `if (OperatingSystem.IsXXX[VersionAtLeast](...))`; runtime/IL linker may turn it into no-op or compile it as runtime check

Swift equivalents:
1. `@available` attribute
2. Same as Clang
3. Swift compiler will emit warning/error
4. `if #available(iOS 15, *)` or `guard #available(iOS 15, *)`
5. Same linker as with Clang (ld64 / ld-prime), same mechanism

**Comment by vcsjones** [comment]
For NativeAOT failures, the "osx" restriction should be changed to `$(_IsApplePlatform)' == 'true'`

here

https://github.com/dotnet/runtime/blob/4f96b8f541a62245eb5ca4b643363785cc0d58bf/src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.Unix.targets#L180

and here

https://github.com/dotnet/runtime/blob/4f96b8f541a62245eb5ca4b643363785cc0d58bf/src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.Unix.targets#L191-L192

File: src/libraries/System.Security.Cryptography/tests/AesCcmTests.cs (hunk 3 above)
Comment (akoeplinger) [reply src/libraries/System.Security.Cryptography/tests/AesCcmTests.cs:15]: Correct, but they are not running because of the `[ConditionalClass(typeof(AesCcm), nameof(AesCcm.IsSupported))]`

I removed the ActiveIssue since it was closed and the tests are correctly skipped.

File: src/libraries/Common/tests/System/Security/Cryptography/AlgorithmImplementations/ECDsa/ECDsaTests.netcoreapp.cs (hunk 2 above)
Comment (akoeplinger) [reply src/libraries/Common/tests/System/Security/Cryptography/AlgorithmImplementations/ECDsa/ECDsaTests.netcoreapp.cs:191]: Yes it passes.

File: src/native/libs/System.Security.Cryptography.Native.Apple/extra_libs.cmake (hunk 4 above)
Comment (akoeplinger) [reply src/native/libs/System.Security.Cryptography.Native.Apple/extra_libs.cmake:8]: Yes I checked and the compiler is automatically weakly-linking against CryptoKit.

**Comment by akoeplinger** [comment]
> For NativeAOT failures, the "osx" restriction should be changed to $(_IsApplePlatform)' == 'true'

Interesting, I explicitly checked this locally and it worked without that change for me.

**Comment by vcsjones** [comment]
@akoeplinger oh and probably https://github.com/dotnet/runtime/blob/4f96b8f541a62245eb5ca4b643363785cc0d58bf/src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.Unix.targets#L217

File: src/libraries/System.Security.Cryptography/tests/AesGcmTests.cs [hunk 8]
```diff
@@ -203,7 +196,7 @@ public static void TwoEncryptionsAndDecryptionsUsingOneInstance()
             byte[] nonce2 = "8ba10892e8b87d031196bf99".HexToByteArray();
 
             byte[] expectedCiphertext1 = "f1af1fb2d4485cc536d618475d52ff".HexToByteArray();
-            byte[] expectedTag1 = PlatformDetection.IsOSX ?
+            byte[] expectedTag1 = (PlatformDetection.IsOSX || PlatformDetection.UsesMobileAppleCrypto) ?
```
Comment (vcsjones) [inline src/libraries/System.Security.Cryptography/tests/AesGcmTests.cs:199]: Not blocking but we could simplify this to `PlatformDetection.IsApplePlatform`

**Review by vcsjones — COMMENTED** [review]
The changes to the actual crypto parts seem reasonable to me. I will have to defer on the building and linking changes to the appropriate folks though.

**Review by ivanpovazan — APPROVED** [review]
LGTM!

**Comment by akoeplinger** [comment]
/ba-g the one unknown test failure is happening on wasm so can't be related to the changes in this PR.
